File: C++/AbstractFactory/WidgetFactory/widget_factory.cc
#include "widget_factory.h"

std::unique_ptr<Button> WindowsWidgetFactory::create_button() const {
    return std::make_unique<WindowsStyleButton>();
}

std::unique_ptr<Scrollbar> WindowsWidgetFactory::create_scrollbar() const {
    return std::make_unique<WindowsStyleScrollbar>();
}


std::unique_ptr<Button> MacWidgetFactory::create_button() const {
    return std::make_unique<MacStyleButton>();
}

std::unique_ptr<Scrollbar> MacWidgetFactory::create_scrollbar() const {
    return std::make_unique<MacStyleScrollbar>();
}

File: Java/AbstractFactory/WidgetFactory/src/com/ericgtkb/WidgetFactory.java
package com.ericgtkb;

// Factory classes are also often singletons
public interface WidgetFactory {
    // These are factory methods
    Button createButton();
    Scrollbar createScrollbar();
}

class WindowsWidgetFactory implements WidgetFactory {
    @Override
    public Button createButton() {
        return new WindowsStyleButton();
    }

    @Override
    public Scrollbar createScrollbar() {
        return new WindowsStyleScrollbar();
    }
}

class MacWidgetFactory implements WidgetFactory {
    @Override
    public Button createButton() {
        return new MacStyleButton();
    }

    @Override
    public Scrollbar createScrollbar() {
        return new MacStyleScrollbar();
    }
}

